wazo_dird/database/queries/base.py

```diff
@@ -87,3 +101,52 @@ def _get_dird_user(self, session, user_uuid):
             session.flush()
 
         return user
+
+    def _populate_parameters(self, parameters=None):
+        new_params = dict(self.DEFAULTS)
+        if parameters:
+            parameters.setdefault('offset', self.DEFAULTS['offset'])
+            new_params.update(parameters)
+
+        return new_params
+
+    def _apply_search_params(self, rows, order, limit, offset, reverse):
+        if order:
+            try:
+                rows = sorted(
+                    rows,
+                    key=lambda x: unicodedata.normalize('NFKD', x[order]),
+                    reverse=reverse,
+                )
+            except KeyError:
+                raise ValueError(f"order: column '{order}' was not found")
+        elif reverse:
+            rows.reverse()
+
+        if not limit:
+            return rows[offset:]
+        else:
+            return rows[offset : offset + limit]
+
+    def validate_parameters(self, parameters):
+        if int(parameters['offset']) < 0:
+            raise ValueError('offset must be positive number')
+
+        if parameters['limit'] is not None and int(parameters['limit']) <= 0:
+            raise ValueError('limit must be a positive number')
+
+        if parameters['direction'] not in self.SORT_DIRECTIONS.keys():
+            raise ValueError('direction must be asc or desc')
+
+    def _extract_search_params(self, parameters={}):
+        parameters = parameters.copy()
+        parameters = self._populate_parameters(parameters)
```

Comment:
you're copying explicitly, which is good for referential transparency, but `self._populate_parameters` also creates a new dict anyway, so there some superfluous copying here.

if you assume(and guarantee) that your methods avoid mutating their parameters, you don't need to explicitly copy.